firestore/client/field_path.cc

```diff
@@ -0,0 +1,97 @@
+// Copyright 2018 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "firestore/client/field_path.h"
+
+namespace firestore {
+std::regex FieldPath::simple_field_name("[_a-zA-Z][_a-zA-Z0-9]*");
+
+FieldPath::FieldPath(std::vector<std::string> const& parts) : parts_(parts) {
+  for (const auto part : parts) {
+    if (part.empty()) {
+      throw std::invalid_argument("One or more components is empty.");
+    }
+  }
+}
+
+void FieldPath::check_invalid_characters(std::string const& string) {
+  std::array<char, 6> invalid_chars = {"~*/[]"};
+  for (const auto invalid_char : invalid_chars) {
+    if (string.find(invalid_char) != std::string::npos) {
+      throw std::invalid_argument("Invalid characters in string");
+    }
+  }
+}
+
+const std::vector<std::string> FieldPath::split(std::string const& string) {
+  std::vector<std::string> parts;
+  auto copy(string);
+  auto index = copy.find('.');
+  while (index != std::string::npos) {
+    parts.push_back(copy.substr(0, index));
+    copy = copy.substr(index + 1);
+    index = copy.find('.');
+  }
+  parts.push_back(copy);
+  return parts;
+}
+
+const FieldPath FieldPath::from_string(std::string const& string) {
+  check_invalid_characters(string);
+  const auto parts = split(string);
```

Comment:
If you make `split()` return non-const and the `FieldParts()` constructor take a vector by value then you can write:

```C++
FieldPath FieldPath::from_string(std::string const& string) {
  check_invalid_characters(string);
  return FieldPath(split(string));
}
```

and save yourself a number of copies.

Reply:
Would there be compiler optimizations available due to const?